packages/flutter/lib/src/material/overscroll_indicator.dart

```diff
@@ -0,0 +1,214 @@
+/// a translucent arc over the affected edge of the child.
+/// If the OverscrollIndicator's child has more than one Scrollable descendant
+/// the scrollableKey parameter can be used to identify the one to track.
+class OverscrollIndicator extends StatefulWidget {
+  OverscrollIndicator({ Key key, this.scrollableKey, this.child }) : super(key: key) {
+    assert(child != null);
+  }
+
+  final Key scrollableKey;
+  final Widget child;
+
+  @override
+  _OverscrollIndicatorState createState() => new _OverscrollIndicatorState();
+}
+
+class _OverscrollIndicatorState extends State<OverscrollIndicator> {
+  final AnimationController _extentAnimation = new AnimationController(
+    lowerBound: _kMinIndicatorExtent,
+    upperBound: _kMaxIndicatorExtent,
+    duration: _kIndicatorHideDuration
+  );
+
+  Timer _hideTimer;
+  Axis _scrollDirection;
+  double _scrollOffset;
+  double _minScrollOffset;
+  double _maxScrollOffset;
+
+  @override
+  void initState() {
+    super.initState();
+  }
+
+  void _hide() {
+    _hideTimer?.cancel();
+    _hideTimer = null;
+    _extentAnimation.reverse();
+  }
+
+  void _updateState(ScrollableState scrollable) {
+    final ExtentScrollBehavior scrollBehavior = scrollable.scrollBehavior;
+    if (scrollBehavior is! ExtentScrollBehavior)
+      return;
+    _scrollDirection = scrollable.config.scrollDirection;
+    _scrollOffset = scrollable.scrollOffset;
+    _minScrollOffset = scrollBehavior.minScrollOffset;
+    _maxScrollOffset = scrollBehavior.maxScrollOffset;
+  }
+
+  void _onScrollStarted(ScrollableState scrollable) {
+    _updateState(scrollable);
+  }
+
+  void _onScrollUpdated(ScrollableState scrollable) {
+    final double value = scrollable.scrollOffset;
+    if ((value < _minScrollOffset || value > _maxScrollOffset) &&
+        ((value - _scrollOffset).abs() > kPixelScrollTolerance.distance)) {
+      _hideTimer?.cancel();
+      _hideTimer = new Timer(_kIndicatorTimeoutDuration, _hide);
+      // Changing the animation's value causes an implicit setState().
+      _extentAnimation.value = value < _minScrollOffset ? _minScrollOffset - value : value - _maxScrollOffset;
+    }
+    _updateState(scrollable);
+  }
+
+  void _onScrollEnded(ScrollableState scrollable) {
+    _updateState(scrollable);
+    _hide();
+  }
+
+  bool _handleScrollNotification(ScrollNotification notification) {
+    if (config.scrollableKey == null || config.scrollableKey == notification.scrollable.config.key) {
+      final ScrollableState scrollable = notification.scrollable;
+      switch(notification.kind) {
+        case ScrollNotificationKind.started:
+          _onScrollStarted(scrollable);
+          break;
+        case ScrollNotificationKind.updated:
+          _onScrollUpdated(scrollable);
+          break;
+        case ScrollNotificationKind.ended:
+          _onScrollEnded(scrollable);
+          break;
+      }
+    }
+    return false;
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
```

Comment:
We usually call super.dispose at the end so that we don't touch the object after disposing it.